ykrt_internal/src/lib.rs

```diff
@@ -0,0 +1,98 @@
+use std::ffi::{c_void, CString};
+use std::panic::UnwindSafe;
+use std::os::raw::c_char;
+
+use ykcompile::CompiledTrace;
+use yktrace::sir::{SirTrace, SIR};
+use yktrace::tir::TirTrace;
+
+fn handle_panic<T>(f: impl FnOnce() -> T + UnwindSafe) -> T {
+    std::panic::catch_unwind(|| f()).unwrap_or_else(|_| {
+        // It is not safe for panics to cross through the "C" abi.
+        // FIXME use "C unwind" as abi once it is implemented.
+        std::process::abort();
+    })
+}
+
+/// The different ways by which we can collect a trace.
+#[derive(Clone, Copy)]
+#[repr(u8)]
+#[allow(dead_code)]
+enum TracingKind {
+    /// Software tracing via ykrustc.
+    SoftwareTracing = 0,
+    /// Hardware tracing via ykrustc + hwtracer.
+    HardwareTracing = 1,
+}
+
+#[no_mangle]
+unsafe extern "C" fn __ykrt_start_tracing(tracing_kind: u8) -> *mut c_void {
+    handle_panic(|| {
+        let tracing_kind = match tracing_kind {
+            0 => yktrace::TracingKind::SoftwareTracing,
```

Comment:
If we enforce the constraint that the tracing kinds are in sync either side of the boundary, we can simplify this I think.

Reply:
How would we enforce that constraint? Is it just a case of documenting it?

Reply:
Enforcement via documentation, yeah.